common/flatpak-transaction.c

```diff
@@ -4062,6 +4092,272 @@ flatpak_transaction_normalize_ops (FlatpakTransaction *self)
+    return TRUE;
+
+  /* This is the set of runtimes which are no longer needed by something in the
+   * transaction (an uninstall or update). The values are either the relevant
+   * remote name or %NULL. */
+  maybe_unused_runtimes = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, g_free);
+
+  /* This is the set of runtimes being used by an install or update operation
+   * in the transaction. */
+  newly_used_runtimes = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
+
+  /* This is the set of runtimes and apps scheduled for uninstallation or
+   * scheduled to switch runtimes in the transaction, and which are therefore
+   * excluded when calculating used refs. */
+  to_be_excluded = g_ptr_array_new ();
+
+  /* These are the set of operations which may need to be executed before the
+   * uninstall operations added by this function. */
+  run_after_ops = g_ptr_array_new ();
+
+  for (l = priv->ops; l != NULL; l = next)
+    {
+      FlatpakTransactionOperation *op = l->data;
+      FlatpakTransactionOperationType op_type = flatpak_transaction_operation_get_operation_type (op);
+      g_autofree char *runtime_ref = NULL;
+      g_autoptr(GPtrArray) related = NULL;
+      g_autoptr(GBytes) deploy_data = NULL;
+
+      next = l->next;
+
+      if (op->skip)
+        continue;
+
+      g_assert (op_type == FLATPAK_TRANSACTION_OPERATION_UNINSTALL ||
+                op_type == FLATPAK_TRANSACTION_OPERATION_INSTALL ||
+                op_type == FLATPAK_TRANSACTION_OPERATION_INSTALL_BUNDLE ||
+                op_type == FLATPAK_TRANSACTION_OPERATION_UPDATE);
+
+      if (op_type == FLATPAK_TRANSACTION_OPERATION_UNINSTALL)
+        g_ptr_array_add (to_be_excluded, op->ref);
+
+      runtime_ref = op_get_runtime_ref (op);
+      if (runtime_ref != NULL)
+        {
+          g_autofree char *runtime_remote = NULL;
+
+          if (op_type == FLATPAK_TRANSACTION_OPERATION_UNINSTALL)
+            {
+              related = find_related_from_deploy (self, runtime_ref, &runtime_remote);
+            }
+          else
+            {
+              FlatpakTransactionOperation *runtime_op = NULL;
+              runtime_op = flatpak_transaction_get_last_op_for_ref (self, runtime_ref);
+              if (runtime_op != NULL)
+                {
+                  runtime_remote = g_strdup (runtime_op->remote);
+                  if (!op_get_related (self, runtime_op, &related, error))
+                    return FALSE;
+                }
+              else
+                related = find_related_from_deploy (self, runtime_ref, &runtime_remote);
+            }
+
+          if (op_type == FLATPAK_TRANSACTION_OPERATION_UNINSTALL)
+            {
+              g_hash_table_replace (maybe_unused_runtimes, g_strdup (runtime_ref), g_strdup (runtime_remote));
+              g_ptr_array_add (run_after_ops, op);
+            }
+          else
+            g_hash_table_add (newly_used_runtimes, g_strdup (runtime_ref));
+
```

Comment:
Thinking some more on this, I'm not sure the non-uninstall part of this is necessary. If we're uninstalling an app, yes, we want to add its runtime and its related refs to maybe_unused (because this is something completely new we didn't do before).

However, if we are updating or installing an app, aren't we guaranteed that the dependencies of it is either installed, or part of the transaction (from the transaction resolve). Why do we need to compute the dependencies again and add them to newly_added_runtimes? All we need to do is add all the refs from the OPERATION_INSTALL ops to newly_used_runtimes.

In fact, that seems like something we have to do anyway. I.e. if a runtime install is explicitly added to the transaction we don't want to consider it unused, even if *nothing* uses it.

Reply:
> Thinking some more on this, I'm not sure the non-uninstall part of this is necessary. If we're uninstalling an app, yes, we want to add its runtime and its related refs to maybe_unused (because this is something completely new we didn't do before).
> 
> However, if we are updating or installing an app, aren't we guaranteed that the dependencies of it is either installed, or part of the transaction (from the transaction resolve). Why do we need to compute the dependencies again and add them to newly_added_runtimes? All we need to do is add all the refs from the OPERATION_INSTALL ops to newly_used_runtimes.
> 

It's guaranteed that the runtime of the app being installed is itself installed (or part of the txn) but it's not guaranteed that it's used. Maybe someone uninstalled the last app using it with the old FlatpakInstallation API, or maybe if we change the way we determine if something is unused to use the current time as an input (#3799), it wasn't considered unused when its last app was uninstalled, but it would be now. In that case it should still not be uninstalled if something in the current transaction will need it.

> In fact, that seems like something we have to do anyway. I.e. if a runtime install is explicitly added to the transaction we don't want to consider it unused, even if _nothing_ uses it.

Yeah we do need to cover that case, good point.